src/txn/txn_rollback_to_stable.c

```diff
@@ -123,7 +127,64 @@ __txn_abort_newer_insert(
     WT_INSERT *ins;
 
     WT_SKIP_FOREACH (ins, head)
-        __txn_abort_newer_update(session, ins->upd, rollback_timestamp);
+        if (ins->upd != NULL)
+            __txn_abort_newer_update(session, ins->upd, rollback_timestamp);
+}
+
+/*
+ * __txn_abort_row_ondisk_kv --
+ *     Fix the on-disk row K/V version according to the given timestamp.
+ */
+static int
+__txn_abort_row_ondisk_kv(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip, wt_timestamp_t rollback_timestamp)
+{
+    WT_CELL_UNPACK *vpack, _vpack;
+    WT_DECL_RET;
+    WT_PAGE_MODIFY *mod;
+    WT_UPDATE *upd, **upd_entry;
+    size_t size;
+
+    vpack = &_vpack;
+    __wt_row_leaf_value_cell(session, page, rip, NULL, vpack);
+    if (vpack->start_ts > rollback_timestamp) {
+        /* FIXME: Get a visible update from the history store if any. */
+
+        /* No visible updates in the history store, may be a newly inserted key. Remove it. */
+        WT_RET(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_NONE;
+        upd->durable_ts = WT_TS_NONE;
+        upd->start_ts = WT_TS_NONE;
+    } else if (vpack->stop_ts != WT_TS_MAX && vpack->stop_ts > rollback_timestamp) {
+        /*
+         * Clear the remove operation from the key by inserting a TOMBSTONE with infinite
+         * visibility.
+         */
+        WT_RET(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_MAX;
+        upd->durable_ts = WT_TS_MAX;
+        upd->start_ts = WT_TS_MAX;
```

Comment:
This part looks odd to me.

The on-disk value has a stop timestamp that is unstable, so a remove needs to be rolled back.  I would expect to see a standard update inserted containing the value and start time pair from on disk.  Maybe reconciliation will see this future tombstone and update the on-disk stop times, but it seems unnatural.

Put another way, if a value was removed with `commit_ts=100`, we wouldn't expect it to be removed again with `commit_ts=MAX`, so relying on reconciliation to do "the right thing" with a tombstone on top of a removed value seems fragile.

Reply:
Fixed by using the on-disk as a standard update.